NE753: Docs work for External DNS Operator (#41811)
openshift/openshift-docs: networking/external_dns_operator/nw-external-dns-operator-configuration-parameters.adoc

:_content-type: CONCEPT
[id="nw-external-dns-operator-configuration-parameters_{context}"]
= External DNS Operator configuration parameters

The External DNS Operator includes the following configuration parameters:

[cols="3a,8a",options="header"]
|===
|Parameter |Description

|`spec`
|Enables the type of a cloud provider.

[source,yaml]
----
spec:
  provider:
    type: AWS <1>
    aws:
      credentials:
        name: aws-access-key <2>
----
<1> Defines available options such as AWS, GCP and Azure.
<2> Defines a name of the `secret` which contains credentials for your cloud provider.

|`zones`
|Enables your specified AWS zone. If you do not specify zones, `ExternalDNS` discovers all the zones present in your cloud provider account.

[source,yaml]
----
zones:
- "myzoneid" <1>
----

<1> specifies the IDs of hosted zones.

|`domains`
|Enables you to specify AWS zones by their domains. If you do not specify domains, `ExternalDNS` discovers all the zones present in your cloud provider account.

[source,yaml]
----
domains:
- filterType: Include <1>
  matchType: Exact <2>
  name: "myzonedomain1.com" <3>
- filterType: Include
  matchType: Pattern <4>

>   pattern: ".*\\.otherzonedomain.com" <5>

Author: Addressed

----
<1> Instructs `ExternalDNS` to include the domain specified.
<2> Instructs `ExtrnalDNS` that the domain matching has to be exact as opposed to regular expression match.
<3> Defines the exact domain name by which `ExternalDNS` filters.
<4> Sets `regex-domain-filter` flag in `ExternalDNS`. You can limit possible domains by using a Regex filter.
<5> Defines the regex pattern to be used by `ExternalDNS` to filter the domains of the target zones.

|`source`

> |enables the `type` service or route.

Author: Addressed

    matchLabels:
      external-dns.mydomain.org/publish: "yes"
  hostnameAnnotation: "Allow" <5>
  fqdnTemplate:
  - "{{.Name}}.myzonedomain.com" <6>
----
<1> Defines the settings for the source of DNS records.
<2> `ExternalDNS` uses `Service` type as source for creating dns records.
<3> Sets `service-type-filter` flag in `ExternalDNS`. The `serviceType` contains the following fields:
* `default`: `LoadBalancer`
* `expected`: `ClusterIP`
* `NodePort`
* `LoadBalancer`
* `ExternalName`
<4> Ensures that the controller considers only those resources which matches with label filter.
<5> The default value for `hostnameAnnotation` is `Ignore` which instructs `ExternalDNS` to generate DNS records using the templates specified in the field `fqdnTemplates`. When the value is `Allow` the DNS records get generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname` annotation.
<6> External DNS Operator uses a string to generate DNS names from sources that don't define a hostname, or to add a hostname suffix when paired with the fake source.

[source,yaml]
----
source:
  type: OpenShiftRoute <1>
  openshiftRouteOptions:
    routerName: default <2>
----

> <1> Defines an option for route source. External DNS filters routes for creation of dns records in the cloud dns.

Author: Addressed

> <2> If the source is `openshift-route`, then you can pass the ingress controller name.

Reviewer: ```suggestion
<2> If the source is `OpenShiftRoute`, then you can pass the ingress controller name whose canonical hostname will be used as the target for CNAME record.
```

Author: Addressed

> [id="nw-external-dns-operator-configuration-parameters_{context}"]

Author: Addressed

> [id="nw-external-dns-operator-configuration-parameters_{context}"]

Author: Addressed

= External DNS Operator configuration parameters

> Your External DNS Operator offers the following configuration parameters:

Author: Addressed

[cols="3a,8a",options="header"]
|===
|Parameter |Description

|`spec`

> |`spec` parameter enables

Author: Addressed

[source,yaml]
----
spec:
  provider:
    type: AWS <1>
    aws:
      credentials:
        name: aws-access-key <2>
----

> <1> `type` field defines available options such as AWS, GCP and Azure.

Author: Addressed

<2> `name` field defines a name of the `secret` which contains credentials for your cloud provider.

|`zones`

> |`zones` enables your specified AWS zone. When you do not specify zones, `ExternalDNS` discovers all the zones present in your cloud provider account.

Author: Addressed

> |`zones` enables your specified AWS zone. When you do not specify zones, `ExternalDNS` discovers all the zones present in your cloud provider account.

Author: Addressed

[source,yaml]
----
zones:
- "myzoneid" <1>
----

<1> The `zones` specifies the IDs of hosted zones.

|`domains`

> |`domains` field defines the hostnames that the External DNS Operator considers for the record creation to the zone specified in zones section.

Reviewer: ```suggestion
|`domains` enables you to specify AWS zones by their domains. When you do not specify domains, `ExternalDNS` discovers all the zones present in your cloud provider account.
```

Author: Addressed

> |`domains` field defines the hostnames that the External DNS Operator considers for the record creation to the zone specified in zones section.

Author: Addressed

[source,yaml]
----
domains:
- filterType: Include <1>
  matchType: Exact <2>
  name: "myzonedomain1.com" <3>
- filterType: Include
  matchType: Pattern <4>
  pattern: ".*\\.otherzonedomain.com" <5>
----

> <1> The `Include` field checks if the domain is present in the hostname.

Reviewer: ```suggestion
<1> The `Include` field instructs `ExternalDNS` to include the domain specified below.
```

Author: Addressed

> <2> If the source is route then the External DNS Operator checks the `domain` of route hostname. For Example, If the hostname is `test.myzonedomain1.com` then External DNS Operator filters the hostname to create the DNS records in the defined zone ID's present under zone's section.

Reviewer: ```suggestion
<2> The `Exact` field instructs `ExtrnalDNS` that the domain matching has to be exact (as opposed to regular expression match).
```

Author: Addressed

> <2> If the source is route then the External DNS Operator checks the `domain` of route hostname. For Example, If the hostname is `test.myzonedomain1.com` then External DNS Operator filters the hostname to create the DNS records in the defined zone ID's present under zone's section.

Author: Addressed

> <3> The `name` defines the domain name that External DNS Operator filters.

Reviewer: ```suggestion
<3> The `name` defines the exact domain name by which `ExternalDNS` filters.
```

Author: Addressed

> <4> The `matchType` sets `regex-domain-filter` in external DNS. You can limit possible domains and target zones by a Regex filter.

Reviewer: ```suggestion
<4> The `matchType` sets `regex-domain-filter` flag in `ExternalDNS`. You can limit possible domains by a Regex filter.
```

Author: Addressed

> <5> The `pattern` defines the regex pattern to be used by External DNS Operator to filter domains and target zones.

Reviewer: ```suggestion
<5> The `pattern` defines the regex pattern to be used by `ExternalDNS` to filter the domains of the target zones.
```

Author: Addressed

|`source`
|`source`

[source,yaml]
----
source: <1>
  type: Service <2>
  service:
    serviceType:<3>
      - LoadBalancer
      - ClusterIP
  labelFilter: <4>
    matchLabels:
      external-dns.mydomain.org/publish: "yes"
  hostnameAnnotation: "Allow" <5>
  fqdnTemplate:
  - "{{.Name}}.myzonedomain.com" <6>
----

> <1> The `source` fields defines the resource types.

Reviewer: ```suggestion
<1> The `source` field defines the settings for the source of DNS records.
```

Author: Addressed

> <2> External DNS filters source type of `service` for creating dns records.

Reviewer: ```suggestion
<2> `ExternalDNS` uses `Service` type as source for creating dns records.
```

Author: Addressed

> <3> The `serviceType` sets `service-type-filter` in external-dns. The `serviceType` contains `default`: `all`, `expected`: `ClusterIP`, `NodePort`, `LoadBalancer` or `ExternalName` fields.

Reviewer 2: In terms of scanability, consider a bulleted list for the values.

Author: Addressed

> <3> The `serviceType` sets `service-type-filter` in external-dns. The `serviceType` contains `default`: `all`, `expected`: `ClusterIP`, `NodePort`, `LoadBalancer` or `ExternalName` fields.

Reviewer: ```suggestion
<3> The `serviceType` sets `service-type-filter` flag in `ExternalDNS`. The `serviceType` contains `default`: `LoadBalancer`, `expected`: `ClusterIP`, `NodePort`, `LoadBalancer` or `ExternalName` fields.
```

Author: Addressed

<4> The `labelFilter` ensures that the controller considers only those resources which matches with label filter.

> <5> The default value for `hostnameAnnotation` is `Ignore`. When the value is `Allow` the DNS records get generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname`. External DNS Operator generates DNS records using the templates specified in the field `fqdnTemplates`.

Reviewer 2: ```suggestion
<5> The default value for `hostnameAnnotation` is `Ignore`. When the value is `Allow`, the DNS records get generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname`. External DNS Operator generates DNS records using the templates specified in the field `fqdnTemplates`.
```

Author: Addressed

> <5> The default value for `hostnameAnnotation` is `Ignore`. When the value is `Allow` the DNS records get generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname`. External DNS Operator generates DNS records using the templates specified in the field `fqdnTemplates`.

Reviewer: ```suggestion
<5> The default value for `hostnameAnnotation` is `Ignore` which instructs `ExternalDNS` to generate DNS records using the templates specified in the field `fqdnTemplates`. When the value is `Allow` the DNS records get generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname` annotation.
```

Author: Addressed

<6> External DNS Operator uses a string to generate DNS names from sources that don't define a hostname, or to add a hostname suffix when paired with the fake source.

[source,yaml]
----
source:
  type: OpenShiftRoute <1>
  openshiftRouteOptions:

>     routerName: default <2>

Reviewer 3: I think `labelFilter` below is also supported for routes. @miheer Do you think we should add it?
```
  labelFilter:
    matchLabels:
      external-dns.mydomain.org/publish: "yes"
```

Author: @miheer can you please confirm?

Author: I will add the fields

----

<1> The `OpenShiftRoute` field defines an option for route source. External DNS filters routes for creation of dns records in the cloud dns.

> <2> If the source is `openshift-route` then you can pass the ingress controller name. 

Reviewer 2: ```suggestion
<2> If the source is `openshift-route`, then you can pass the ingress controller name. 
```

Author: Addressed
